internal/wasm/interpreter/interpreter.go

```diff
@@ -507,6 +515,12 @@ func (me *moduleEngine) Call(ctx *wasm.ModuleContext, f *wasm.FunctionInstance,
 
 	ce := me.newCallEngine()
 	defer func() {
+		// Default to an exit error if another function closed the module, but allow a better error to override it.
+		if err == nil {
+			if closed := atomic.LoadUint64(&me.closed); closed > 0 {
```

Comment:
@mathetake pretty sure this is correct, just double-check me if you don't mind. I was hoping i could just read me.closed directly, but did like this just in case there's some visibility problem otherwise.

regardless, I think the tests still race and there's a chance things pass this barrier in concurrent scenarios. usually in other tests, I put some sort of ordering condition to make sure something happens after (usually a blocking queue in the tests). I'll sort out the race in tests later, mainly I'm curious your opinion of the impl itself!

Reply:
ps I will be refactoring the lock to be external as there's too much similar between here and jit and easier to test externally

Reply:
this is correct. I did research and added to RATIONALE.md in https://github.com/tetratelabs/wazero/pull/411 as it was hard to find a concrete answer (TL;DR; it isn't concretely answered, but fairly understood)